tools/scxtop/src/bpf/main.bpf.c

```diff
@@ -526,3 +541,56 @@ int BPF_PROG(on_sched_switch, bool preempt, struct task_struct *prev,
 
 	return 0;
 }
+
+SEC("tp_btf/softirq_entry")
+int BPF_PROG(on_softirq_entry, unsigned int nr)
+{
+	struct task_struct *p;
+
+	if (!enable_bpf_events || !should_sample())
+		return 0;
+
+	p = (struct task_struct *)bpf_get_current_task();
+	if (!p)
+		return 0;
+
+	struct __softirq_event event;
+	event.cpu = bpf_get_smp_processor_id();
+	event.start_ts = bpf_ktime_get_ns();
+	event.pid = BPF_CORE_READ(p, pid);
+
+	bpf_map_update_elem(&softirq_events, &zero_int, &event, BPF_ANY);
+
+	return 0;
+}
+
+SEC("tp_btf/softirq_exit")
+int BPF_PROG(on_softirq_exit, unsigned int nr)
+{
+	struct bpf_event *event;
+	struct __softirq_event *softirq_event;
+
+	if (!enable_bpf_events || !should_sample())
+		return 0;
+
+	u64 exit_ts = bpf_ktime_get_ns();
+
+	softirq_event = bpf_map_lookup_elem(&softirq_events, &zero_int);
+	if (!softirq_event)
+		return 0;
+
+	if (!(event = try_reserve_event()))
+		return -ENOENT;
+
+	event->type = SOFTIRQ;
+	event->cpu = softirq_event->cpu;
+	event->ts = exit_ts;
+	event->event.softirq.pid = softirq_event->pid;
+	event->event.softirq.entry_ts = softirq_event->start_ts;
```

Comment:
Do we need `cpu` and `pid` from `softirq_event`, or could it simply be the entry timestamp? Wondering if it'd be kinder to the `irq` to do as much work as possible on `exit` rather than `entry`.